extraction/src/main/scala/org/broadinstitute/monster/encode/extraction/ExtractionPipelineBuilder.scala

```diff
@@ -21,224 +18,232 @@ import scala.concurrent.Future
-          s"${args.outputDir}/${entityToExtract.entryName}"
-        )
-        out
+      val description =
+        s"Query ${targetEntityType.entryName} data using ${sourceEntityType.entryName} objects"
+
+      val out = sourceEntities.transform(description) { entities =>
+        val joinValues =
+          entities.flatMap(_.tryRead[Array[String]](sourceField).getOrElse(Array.empty))
+        val queryBatches = groupValues(batchSize, joinValues.map(targetField -> _))
+        getEntities(targetEntityType, queryBatches).distinctBy(_.read[String]("@id"))
       }
+      writeJsonLists(
+        out,
+        s"${targetEntityType.entryName}",
+        s"${args.outputDir}/${targetEntityType.entryName}"
+      )
+      out
     }
 
     // biosamples are the first one and follow a different pattern, so we don't use the generic method
     val biosamples = ctx
+      .withName("Inject initial query")
       .parallelize(List(args.initialQuery))
-      .transform(s"Extract ${EncodeEntity.Biosample} data") { rawData =>
-        val biosamples = getEntities(EncodeEntity.Biosample)(rawData)
-        writeJsonLists(
-          biosamples,
-          s"${EncodeEntity.Biosample}",
-          s"${args.outputDir}/${EncodeEntity.Biosample.entryName}"
-        )
-        biosamples
+      .transform(s"Extract ${EncodeEntity.Biosample} data") { initialQuery =>
+        // No need to de-duplicate here, there's only one query batch and
+        // ENCODE won't return the same object twice in one query.
+        getEntities(EncodeEntity.Biosample, initialQuery)
       }
+    writeJsonLists(
+      biosamples,
+      s"${EncodeEntity.Biosample}",
+      s"${args.outputDir}/${EncodeEntity.Biosample.entryName}"
+    )
 
     // don't need to use donors apart from storing them, so we don't assign an output here
-    extractLinkedEntities(EncodeEntity.Donor, "donor", biosamples)
+    extractLinkedEntities(
+      sourceEntityType = EncodeEntity.Biosample,
+      sourceField = "donor",
+      sourceEntities = biosamples,
+      targetEntityType = EncodeEntity.Donor,
+      targetField = "@id"
+    )
 
     val libraries = extractLinkedEntities(
-      entityToExtract = EncodeEntity.Library,
-      matchingField = "accession",
-      linkingEntities = biosamples,
-      linkedField = "biosample.accession"
+      sourceEntityType = EncodeEntity.Biosample,
+      sourceField = "accession",
+      sourceEntities = biosamples,
+      targetEntityType = EncodeEntity.Library,
+      targetField = "biosample.accession"
     )
 
-    val replicates = extractLinkedEntities(
-      entityToExtract = EncodeEntity.Replicate,
-      matchingField = "accession",
-      linkingEntities = libraries,
-      linkedField = "library.accession"
+    val files = extractLinkedEntities(
+      sourceEntityType = EncodeEntity.Library,
+      sourceField = "@id",
+      sourceEntities = libraries,
+      targetEntityType = EncodeEntity.File,
+      targetField = "replicate_libraries"
     )
```

Comment:
Querying files using experiments was bad either way we tried it.
* Querying on `file.dataset = experiment.id` resulted in huge response payloads
* Querying on `file.id = experiment.files` resulted in too many individual HTTP requests

I found this `replicate_libraries` field on the file objects. It's a roll-up of the file's direct libraries (for sequence files) and the libraries of all `derived_from` files. Running this code in Dataflow finished in a few minutes vs. 30+ minutes in the previous implementation.